apisix/plugins/limit-conn/limit-conn-redis.lua

```diff
@@ -0,0 +1,139 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local redis     = require("apisix.utils.redis")
+local core          = require("apisix.core")
+local assert        = assert
+local math          = require "math"
+local floor         = math.floor
+local ngx_timer_at  = ngx.timer.at
+
+local setmetatable  = setmetatable
+
+
+local _M = {version = 0.1}
+
+
+local mt = {
+    __index = _M
+}
+
+function _M.new(plugin_name, conf, max, burst, default_conn_delay)
+
+    local self = {
+        conf = conf,
+        plugin_name = plugin_name,
+        burst = burst,
+        max = max + 0,    -- just to ensure the param is good
+        unit_delay = default_conn_delay,
+    }
+    return setmetatable(self, mt)
+end
+
+
+function _M.incoming(self, key, commit)
+    local max = self.max
+
+    local conf = self.conf
+    local red, err = redis.new(conf)
```

Comment:
move this to `_M.new()`

Reply:
This cannot be placed in new because you cannot ensure whether it times out.

Reply:
what do you mean?

Reply:
> what do you mean?

Because Redis uses a cosocket for its operations, it is susceptible to timeouts. However, RedisCluster is built on resty.redis. It establishes a connection every time do_cmd is executed. Therefore, we need to establish a Redis connection every time there is incoming or leaving activity to prevent cosocket timeouts.